agent/main.go

```diff
@@ -285,6 +304,26 @@ func mainImpl() {
 		)
 	}
 
+	{
```

Comment:
Nit: `init` is ~300 lines long, which makes it hard (at least for me) to understand what it does at a glance.
It feels like it could be splitted into smaller functions to separate the "what" from the "how"?

Reply:
Yes agreed, but for next PR I think I won't have time now, sadly.